govcd/nsxt_distributed_firewall_test.go

```diff
@@ -283,3 +283,130 @@ func dumpDistributedFirewallRulesToScreen(rules []*types.DistributedFirewallRule
+	}
+	skipNoNsxtConfiguration(vcd, check)
+	skipOpenApiEndpointTest(vcd, check, types.OpenApiPathVersion1_0_0+types.OpenApiEndpointEdgeGateways)
+
+	adminOrg, err := vcd.client.GetAdminOrgByName(vcd.config.VCD.Org)
+	check.Assert(adminOrg, NotNil)
+	check.Assert(err, IsNil)
+
+	nsxtExternalNetwork, err := GetExternalNetworkV2ByName(vcd.client, vcd.config.VCD.Nsxt.ExternalNetwork)
+	check.Assert(nsxtExternalNetwork, NotNil)
+	check.Assert(err, IsNil)
+
+	vdc, vdcGroup := test_CreateVdcGroup(check, adminOrg, vcd)
+	check.Assert(vdc, NotNil)
+	check.Assert(vdcGroup, NotNil)
+
+	fmt.Println("# Running Distributed Firewall tests for single Rule")
+	test_NsxtDistributedFirewallRule(vcd, check, vdcGroup.VdcGroup.Id, vcd.client, vdc)
+
+	// Cleanup
+	err = vdcGroup.Delete()
+	check.Assert(err, IsNil)
+	err = vdc.DeleteWait(true, true)
+	check.Assert(err, IsNil)
+}
+
+func test_NsxtDistributedFirewallRule(vcd *TestVCD, check *C, vdcGroupId string, vcdClient *VCDClient, vdc *Vdc) {
+	adminOrg, err := vcdClient.GetAdminOrgByName(vcd.config.VCD.Org)
+	check.Assert(adminOrg, NotNil)
+	check.Assert(err, IsNil)
+
+	vdcGroup, err := adminOrg.GetVdcGroupById(vdcGroupId)
+	check.Assert(err, IsNil)
+
+	_, err = vdcGroup.ActivateDfw()
+	check.Assert(err, IsNil)
+
+	// Prep firewall rule sample to operate with
+	randomizedFwRuleDefs, ipSet, secGroup := createDistributedFirewallDefinitions(check, vcd, vdcGroup.VdcGroup.Id, vcdClient, vdc)
+	// defer cleanup function in case something goes wrong
+	defer func() {
+		dfw, err := vdcGroup.GetDistributedFirewall()
+		check.Assert(err, IsNil)
+		err = dfw.DeleteAllRules()
+		check.Assert(err, IsNil)
+		_, err = vdcGroup.DisableDefaultPolicy()
+		check.Assert(err, IsNil)
+		err = ipSet.Delete()
+		check.Assert(err, IsNil)
+		err = secGroup.Delete()
+		check.Assert(err, IsNil)
+	}()
+
+	randomizedFwRuleSubSet := randomizedFwRuleDefs[0:5] // taking only first 10 rules to limit time of testing
+
+	// removing default firewall rule which is created by VCD when vdcGroup.ActivateDfw() is executed
+	err = vdcGroup.DeleteAllDistributedFirewallRules()
+	check.Assert(err, IsNil)
+
+	// Adding firewal rules one by one and checking that each of them is
+	testDistributedFirewallRuleSequence(vcd, check, randomizedFwRuleSubSet, vdcGroup, false)
+	testDistributedFirewallRuleSequence(vcd, check, randomizedFwRuleSubSet, vdcGroup, true)
+}
+
+// testDistributedFirewallRuleSequence tests the following:
+// * create firewall rules one one by one
+// * check that the order of firewall rules is the same as requested (or exactly reverse if
+// reverseOrder=true)
+// * check that all IDs of created firewall rules persisted during further updates (means that no
+// firewall rules were recreated during addition of new ones)
+func testDistributedFirewallRuleSequence(vcd *TestVCD, check *C, randomizedFwRuleSubSet []*types.DistributedFirewallRule, vdcGroup *VdcGroup, reverseOrder bool) {
```

Comment:
Would it be possible to add a test to hit parallel creation case with the goal to corrupt data? This ask is coming from my comment above: if we do add lock to the public function, then this test should never corrupt data. In a sense, it's a TDD: if it's doable at all, first you could write the test, hit the data corruption and afterwards add the lock and see that there's no more data corruption happening.

CC: @dataclouder

Reply:
I can add a test that replicates a problem - the outcome wouldn't be exactly predictable (some rules could make it, others wouldn't - even depending on latency), but the fact is that it wouldn't be as predicted. But OTOH - I am not sure if we want to perform locking itself - https://github.com/vmware/go-vcloud-director/pull/587#discussion_r1252894679